OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs

```diff
@@ -51,27 +52,26 @@ public void AttackMove(CPos cell, int closeEnough = 0)
 			"and the actor will wait for `wait` ticks at each waypoint.")]
 		public void Patrol(CPos[] waypoints, bool loop = true, int wait = 0)
 		{
-			foreach (var wpt in waypoints)
-			{
-				Self.QueueActivity(new AttackMoveActivity(Self, () => move.MoveTo(wpt, 2)));
-				Self.QueueActivity(new Wait(wait));
-			}
-
-			if (loop)
-				Self.QueueActivity(new CallFunc(() => Patrol(waypoints, loop, wait)));
+			Self.QueueActivity(new Patrol(Self, waypoints, loop, wait));
 		}
 
 		[ScriptActorPropertyActivity]
 		[Desc("Patrol along a set of given waypoints until a condition becomes true. " +
 			"The actor will wait for `wait` ticks at each waypoint.")]
 		public void PatrolUntil(CPos[] waypoints, LuaFunction func, int wait = 0)
 		{
-			Patrol(waypoints, false, wait);
+			var lf = func.CopyReference() as LuaFunction;
+			var f = new Func<bool>(() =>
+			{
+				return lf.Call(Self.ToLuaValue(Context)).First().ToBoolean();
+			});
 
-			var repeat = func.Call(Self.ToLuaValue(Context)).First().ToBoolean();
-			if (repeat)
-				using (var f = func.CopyReference() as LuaFunction)
-					Self.QueueActivity(new CallFunc(() => PatrolUntil(waypoints, f, wait)));
+			Self.QueueActivity(new Patrol(Self, waypoints, f, wait));
+			Self.QueueActivity(new CallFunc(() =>
+			{
+				if (lf != null)
+					lf.Dispose();
+			}));
```

Comment:
I'm not sure if this works. If the actor is stopped in anyway this activity is thrown away and `lf` never disposed.
Can we not keep `using`?
```
			using (var lf = func.CopyReference() as LuaFunction)
			{
				var f = new Func<bool>(() =>lf.Call(Self.ToLuaValue(Context)).First().ToBoolean());
				Self.QueueActivity(new Patrol(Self, waypoints, f, wait));
			}
```

Reply:
> Can we not keep `using`?

Maybe. Unfortunately, your suggestion crashes with an NRE. I suppose `using` is a bit too eager to dispose `lf` in this situation. I solved it by pulling this into the activity itself.